### PR TITLE
RBD image live in migration from s3 object

### PR DESCRIPTION
**Pull request by VasishtaShastry**
# Description
RBD image live in migration from s3 object

Test cases covered :
CEPH-83574092 - verify snapshot downloaded creates a local RBD image

Test Case Flow :
1. Create images and take snapshots, export snapshots.
2. Backup exported snapshots as rbd images.
3. Mention s3 object as source for live migration preparation.
4. Execute and commit migration and verify.

File changes:
modified:   conf/pacific/rbd/tier-0_rbd.yaml
and conf/quincy/rbd/tier-0_rbd.yaml
Added rgw node required for the suite.

modified:   suites/pacific/rbd/tier-2_rbd_regression.yaml
and suites/quincy/rbd/tier-2_rbd_regression.yaml
Added tests required for testcase execution

modified:   tests/rbd/rbd_utils.py
Modified doc strings of a module with ore details
Added capability to mention source for migration prepare method.

new file:   tests/rbd/test_rbd_migration_image_s3_object.py
Module to handle the testcase by executing steps and initiate boto script to
upload exported image located in ceph-qe-scripts.

Signed-off-by: Vasishta <vashastr@redhat.com>

April 11th
Added more doc

Addressed review comments and tox errors
Log links - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-ISE251
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-9NGUK0/Image_migration_from_s3_source_0.log
Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/rbd/rbd_utils.py [hunk 1]
```diff
@@ -204,6 +204,19 @@ def check_pool_exists(self, pool_name: str) -> bool:
             )
             return False
 
+    def export(self, imagespec, path):
```
Comment (rkhambad) [inline tests/rbd/rbd_utils.py:207]: This method is already added in rbd_uitls in one of the latest PR merge https://github.com/red-hat-storage/cephci/pull/2393 , Please rebase your branch.

**Review by rkhambad — CHANGES_REQUESTED** [review]

File: suites/pacific/rbd/tier-2_rbd_regression.yaml [hunk 2]
```diff
@@ -174,6 +182,28 @@ tests:
...
+        ec_pool_config:
+          pool: rbd_pool
+          data_pool: rbd_ec_pool
+          image: rbd_image
+          size: 100M
+        rep_pool_config:
+          pool: rbd_pool
+          image: rbd_rep_image
+          size: 100M
+
```
Comment (rkhambad) [inline suites/pacific/rbd/tier-2_rbd_regression.yaml:206]: Polarion ID is missing for suite

File: tests/rbd/test_rbd_migration_image_s3_object.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        rbd_obj.exec_cmd(cmd=f'echo "{source_json}" >> source.json')
+
+        rbd_obj.migration_prepare("source.json", f"{each_config[0:2]}_import")
+        rbd_obj.migration_action("execute", f"{each_config[0:2]}_import")
+        rbd_obj.migration_action("commit", f"{each_config[0:2]}_import")
+        verify_migration_commit(
+            rbd_obj,
+            f"{kw['config'][each_config]['pool']}",
+            f"{kw['config'][each_config]['image']}",
+        )
```
Comment (rkhambad) [inline tests/rbd/test_rbd_migration_image_s3_object.py:93]: Please add pool cleanup

File: tests/rbd/rbd_utils.py (hunk 1 above)
Comment (VasishtaShastry) [reply tests/rbd/rbd_utils.py]: Addressed, thank you

File: tests/rbd/test_rbd_migration_image_s3_object.py (hunk 3 above)
Comment (VasishtaShastry) [reply tests/rbd/test_rbd_migration_image_s3_object.py:106]: Added.

File: suites/pacific/rbd/tier-2_rbd_regression.yaml (hunk 2 above)
Comment (VasishtaShastry) [reply suites/pacific/rbd/tier-2_rbd_regression.yaml]: Added.

File: suites/pacific/rbd/tier-2_rbd_regression.yaml [hunk 4]
```diff
@@ -222,6 +229,28 @@ tests:
...
       polarion-id: CEPH-83573327
 
+  - test:
+      desc: Setting up RGW user for s3 image migration test
+      module: exec.py
+      name: Setup rgw s3 user
+      config:
+        commands:
+          - "radosgw-admin user create --uid=test_rbd --display_name='test_rbd' --access-key test_rbd --secret test_rbd"
+  - test:
```
Comment (sunilangadi2) [inline suites/pacific/rbd/tier-2_rbd_regression.yaml:239]: please add line spacing for each test.

File: tests/rbd/test_rbd_migration_image_s3_object.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        rbd_obj.migration_prepare("source.json", destination_spec)
+        rbd_obj.migration_action("execute", destination_spec)
+        rbd_obj.migration_action("commit", destination_spec)
+        verify_migration_commit(
+            rbd_obj,
+            f"{kw['config'][each_config]['pool']}",
+            f"{kw['config'][each_config]['image']}",
+        )
+
+    cleanup(rbd_obj, kw["config"])
```
Comment (sunilangadi2) [inline tests/rbd/test_rbd_migration_image_s3_object.py:108]: Please add cleanup in `Finally` clause

File: conf/pacific/rbd/tier-0_rbd.yaml [hunk 6]
```diff
@@ -28,6 +28,7 @@ globals:
         role:
           - mds
           - osd
+          - rgw
```
Comment (sunilangadi2) [inline conf/pacific/rbd/tier-0_rbd.yaml:31]: Please cover this daemon and test in quincy as well.

**Comment by rkhambad** [comment]
please resolve the tox issue

File: tests/rbd/rbd_utils.py [hunk 7]
```diff
@@ -647,8 +648,13 @@ def migration_prepare(self, src_spec, dest_spec):
...
             dest_spec: Target image spec TARGET_POOL_NAME/SOURCE_IMAGE_NAME
         """
-        log.info("Starting prepare Live migration of image ")
-        return self.exec_cmd(cmd=f"rbd migration prepare {src_spec} {dest_spec}")
+        log.info("Starting prepare Live migration of image")
+        command = "rbd migration prepare "
+        if ".json" not in src_spec:
+            command += f"{src_spec} {dest_spec}"
+        else:
+            command += f"{dest_spec} --import-only --source-spec-path {src_spec}"
```
Comment (rkhambad) [inline tests/rbd/rbd_utils.py:656]: please add info about --import-only in method doc

File: tests/rbd/test_rbd_migration_image_s3_object.py (hunk 5 above)
Comment (VasishtaShastry) [reply tests/rbd/test_rbd_migration_image_s3_object.py]: Done.

File: conf/pacific/rbd/tier-0_rbd.yaml (hunk 6 above)
Comment (VasishtaShastry) [reply conf/pacific/rbd/tier-0_rbd.yaml:31]: Done.

File: tests/rbd/rbd_utils.py (hunk 7 above)
Comment (VasishtaShastry) [reply tests/rbd/rbd_utils.py]: Added doc about method input

File: source.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
+    "type": "raw",
+    "stream": {
+        "type": "s3",
+        "url": "http://10.0.208.243:80/rbd/re",
+        "access_key": "test_rbd",
+        "secret_key": "test_rbd"
+    }
+}
```
Comment (manasagowri) [inline source.json:9]: Please add a newline at the end of file

Comment (VasishtaShastry) [reply source.json]: Inclusion of that file was a mishap
Have removed it
sorry

File: suites/quincy/rbd/tier-2_rbd_regression.yaml [hunk 9]
```diff
@@ -255,6 +263,30 @@ tests:
...
+      desc: Export an image to s3 source and reinstall as new image using migration
+      module: test_rbd_migration_image_s3_object.py
+      polarion-id: CEPH-83574092
+      config:
+        ec_pool_config:
+          pool: rbd_pool
+          data_pool: rbd_ec_pool
+          image: rbd_image
+          size: 100M
+        rep_pool_config:
```
Comment (rkhambad) [inline suites/quincy/rbd/tier-2_rbd_regression.yaml:285]: Does it cover Replication pool also, I did not see any execution for rep_image.?
In logs attached i can see only ec pool image migration.
log link checked http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LJHAYF/ 
if there is any other log ink please correct me.

**Review by rkhambad — COMMENTED** [review]
Also please modify the Polarion test steps, its does not look like actual steps.

**Review by sunilangadi2 — APPROVED** [review]
My Changes LGTM

**Review by rkhambad — APPROVED** [review]
Addressed all changes. LGTM

File: source.json (hunk 8 above)
Comment (VasishtaShastry) [reply source.json]: Marking as resolved as the file has been removed.
Please feel free to revert if I missed something.

File: suites/quincy/rbd/tier-2_rbd_regression.yaml (hunk 9 above)
Comment (VasishtaShastry) [reply suites/quincy/rbd/tier-2_rbd_regression.yaml:285]: Marking as resolved as have received approval from Rajendra

**Comment by Manohar-Murthy** [comment]
team review is done.

**Pull request merged**